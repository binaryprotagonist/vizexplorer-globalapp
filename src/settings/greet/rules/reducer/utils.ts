import {
  GreetRuleFragment,
  PdGreetAssignmentType,
  PdGuestInteractionType
} from "generated-graphql";
import {
  AssignTo,
  GreetDraftMetricTrigger,
  GreetRuleDraft,
  OverflowAssignTo,
  OverflowGreetAssignmentType
} from "./types";

export function isAssignmentGroupComplete(group: AssignTo | OverflowAssignTo): boolean {
  if (group.assignmentToType === PdGreetAssignmentType.SpecificUserGroup) {
    return !!group.userGroupId;
  }

  return true;
}

export function isGuestAndAssignmentTypeCompatible(
  assignmentType: OverflowGreetAssignmentType,
  guestType: PdGuestInteractionType
): boolean {
  if (guestType === PdGuestInteractionType.Coded || assignmentType === "None") {
    return true;
  }

  return (
    assignmentType === PdGreetAssignmentType.AllUsers ||
    assignmentType === PdGreetAssignmentType.SpecificUserGroup
  );
}

export function assignmentTypeAllowsOverflow(
  assignmentType: OverflowGreetAssignmentType
): boolean {
  return assignmentType !== "None" && assignmentType !== PdGreetAssignmentType.AllUsers;
}

export function isWeightValid(value: number | null): boolean {
  return value !== null && value >= 0 && value <= 100;
}

export function isMetricValid(metric: GreetDraftMetricTrigger): boolean {
  if (!metric.value) return false;
  if (metric.value[0] === null && metric.value[1] === null) return false;
  if (metric.value[0] === null && metric.value[1] !== null) return true;
  if (metric.value[0] !== null && metric.value[1] === null) return true;
  if (metric.value[0]! <= metric.value[1]!) return true;
  return false;
}

export function isNameTaken(
  existingName: string,
  newName: string,
  rules: GreetRuleFragment[]
): boolean {
  const newNameLower = newName.trim().toLowerCase();
  const existingNameLower = existingName.trim().toLowerCase();
  if (existingNameLower === newNameLower) return false;

  return rules.some((rule) => rule.name.trim().toLowerCase() === newNameLower);
}

function isRequiredFieldsComplete(rule: GreetRuleDraft): boolean {
  return (
    !!rule.name &&
    !!rule.siteId &&
    !!rule.specialTriggers.sections?.length &&
    !!rule.specialTriggers.guestType &&
    !!rule.specialTriggers.tiers?.length &&
    !!rule.specialTriggers.daysOfWeeks?.length &&
    !!rule.assignment.assignTo?.assignmentToType
  );
}

function isPartialCompleteMetric(metric: GreetDraftMetricTrigger): boolean {
  return !!metric.code && !isMetricValid(metric);
}

export function isRuleComplete(
  rule: GreetRuleDraft,
  rules: GreetRuleFragment[],
  initialRuleName: string
): boolean {
  if (!isRequiredFieldsComplete(rule)) return false;
  if (isNameTaken(initialRuleName, rule.name, rules)) return false;
  if (rule.metricTriggers.some(isPartialCompleteMetric)) return false;
  if (!isWeightValid(rule.assignment.weight)) return false;
  // Validated non-null in `isRequiredFieldsComplete`
  if (!isAssignmentGroupComplete(rule.assignment.assignTo!)) return false;
  const overflow1 = rule.assignment.overflowGroup1;
  if (overflow1 && !isAssignmentGroupComplete(overflow1)) return false;
  const overflow2 = rule.assignment.overflowGroup2;
  if (overflow2 && !isAssignmentGroupComplete(overflow2)) return false;

  return true;
}
