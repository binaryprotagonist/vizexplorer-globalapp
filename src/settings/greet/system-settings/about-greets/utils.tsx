import { Typography } from "@vizexplorer/global-ui-v2";
import { AboutGreetsSettingId } from "./types";

export function aboutGreetsSettingTitle(settingId: AboutGreetsSettingId): string {
  switch (settingId) {
    case "assign-by-priority-score":
      return "Assign by priority score";
    case "suppression-days-after-completion":
      return "Suppression days after completion";
    case "queue-inactive-greet-timeout":
      return "Queue inactive greet timeout";
    case "greet-reassignment-timeout":
      return "Greet reassignment timeout";
    case "show-guests-active-actions":
      return "Show guest active actions";
  }
}

export function aboutGreetsSettingHelp(settingId: AboutGreetsSettingId): JSX.Element {
  switch (settingId) {
    case "assign-by-priority-score":
      return (
        <>
          <Typography variant={"label"} color={"white"} fontWeight={600}>
            Priority score
          </Typography>
          <Typography variant={"label"} color={"white"} marginY={1}>
            The system will assign actions based on their priority scores, which are
            calculated using the assignment weight, tier score, and time factors.
          </Typography>
          <Typography variant={"label"} color={"white"}>
            When disabled, actions are assigned sequentially from the top of the queue to
            the bottom.
          </Typography>
        </>
      );
    case "suppression-days-after-completion":
      return (
        <>
          Duration during which the system will not generate new greets for uncoded and
          coded guests after a host completes a greet.
        </>
      );
    case "queue-inactive-greet-timeout":
      return (
        <>
          Number of minutes a greet can remain in the queue before it is automatically
          removed.
        </>
      );
    case "greet-reassignment-timeout":
      return (
        <>
          Number of minutes from greet assignment before it is considered a missed greet
          and automatically reassigned to another person.
        </>
      );
    case "show-guests-active-actions":
      return (
        <>
          Display the number of other active actions associated with that guest in the
          action list. This helps provide visibility into the workload of each guest and
          their ongoing actions.
        </>
      );
  }
}
