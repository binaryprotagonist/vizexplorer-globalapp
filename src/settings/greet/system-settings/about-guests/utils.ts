import { AboutGuestsSettingId } from "./types";

export function aboutGuestsSettingTitle(settingId: AboutGuestsSettingId): string {
  switch (settingId) {
    case "report-banned-guest":
      return "Report banned guest";
  }
}

export function aboutGuestsSettingHelp(settingId: AboutGuestsSettingId): string {
  switch (settingId) {
    case "report-banned-guest":
      return "If a Banned Guest cards in, report it to someone.";
  }
}
