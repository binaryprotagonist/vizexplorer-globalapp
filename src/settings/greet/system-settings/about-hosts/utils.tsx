import { AboutHostsSettingId } from "./types";

export function aboutHostsSettingTitle(settingId: AboutHostsSettingId): string {
  switch (settingId) {
    case "enable-section-for-greet":
      return "Enable sections for greets assignment";
    case "allow-suppression-without-completion":
      return "Allow suppression without completion";
    case "max-assignment-per-host":
      return "Max assignments per host";
    case "max-missed-greets":
      return "Max missed greets";
  }
}

export function aboutHostsSettingHelp(settingId: AboutHostsSettingId): JSX.Element {
  switch (settingId) {
    case "enable-section-for-greet":
      return (
        <>
          Use sections when determining greet assignment. When enabled, host will be
          required to choose their sections. If disabled, the section picker will be
          turned off.
        </>
      );

    case "allow-suppression-without-completion":
      return (
        <>
          Allow Hosts to suppress a greet without completing it.
          <br />A Suppress button will appear in the greet cards.
        </>
      );

    case "max-assignment-per-host":
      return <>Number of greets a host can be assigned.</>;

    case "max-missed-greets":
      return (
        <>
          Number of consecutive missed greets that will trigger an automatic status change
          to unavailable.
        </>
      );
  }
}
