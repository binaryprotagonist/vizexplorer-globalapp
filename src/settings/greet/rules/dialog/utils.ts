import {
  GreetRuleFragment,
  GreetRuleGroupAssignmentFragment,
  GroupAssignmentCreateInput,
  PdGreetAssignmentType,
  PdGreetRuleCreateInput,
  PdGreetRuleMetricTriggerInput,
  PdGreetRuleSpecialTriggerCreateInput,
  PdGreetRuleSpecialTriggerValue,
  PdGreetRuleTriggerType,
  PdGreetRuleUpdateInput,
  PdGuestInteractionType
} from "generated-graphql";
import { GreetRuleDraft, assignmentTypeAllowsOverflow, isMetricValid } from "../reducer";
import {
  AssignTo,
  DraftMultiSelectValue,
  GreetDraftMetricTrigger,
  GreetDraftSiteTriggers,
  OverflowAssignTo
} from "../reducer/types";
import { GreetRuleTrigger } from "../types";
import { isMetricValue } from "../utils";

export function draftRuleToRuleCreateInput(rule: GreetRuleDraft): PdGreetRuleCreateInput {
  const assignTo = draftAssignmentAsAssignmentInput(rule.assignment.assignTo!)!;
  const overflowGroup1 = rule.assignment.overflowGroup1
    ? draftAssignmentAsAssignmentInput(rule.assignment.overflowGroup1)
    : null;
  const overflowGroup2 = rule.assignment.overflowGroup2
    ? draftAssignmentAsAssignmentInput(rule.assignment.overflowGroup2)
    : null;

  return {
    name: rule.name,
    siteId: rule.siteId!,
    isEnabled: rule.isEnabled,
    isIgnoreSuppression: rule.isIgnoreSuppression,
    specialTriggers: draftSpecialAsSpecialInput(rule.specialTriggers),
    metricTriggers: draftMetricsAsMetricInput(rule.metricTriggers),
    assignment: {
      weight: rule.assignment.weight!,
      assignTo,
      ...(overflowGroup1 && {
        overflowAssignment1: overflowGroup1
      }),
      ...(overflowGroup2 && {
        overflowAssignment2: overflowGroup2
      })
    }
  };
}

export function draftRuleToUpdateInput(rule: GreetRuleDraft): PdGreetRuleUpdateInput {
  return {
    ...draftRuleToRuleCreateInput(rule),
    id: rule.id!
  };
}

function specialTriggerKeyAsType(
  key: keyof GreetDraftSiteTriggers
): PdGreetRuleTriggerType {
  switch (key) {
    case "daysOfWeeks":
      return PdGreetRuleTriggerType.DaysOfWeeks;
    case "guestType":
      return PdGreetRuleTriggerType.GuestType;
    case "sections":
      return PdGreetRuleTriggerType.Section;
    case "tiers":
      return PdGreetRuleTriggerType.Tier;
  }
}

function draftSpecialAsSpecialInput(
  specials: GreetDraftSiteTriggers
): PdGreetRuleSpecialTriggerCreateInput[] {
  return Object.entries(specials)
    .filter(([_, value]) => value !== "__ALL__")
    .map<PdGreetRuleSpecialTriggerCreateInput>(([key, value]) => ({
      type: specialTriggerKeyAsType(key as keyof GreetDraftSiteTriggers),
      value: value!
    }));
}

function draftMetricsAsMetricInput(
  metrics: GreetDraftMetricTrigger[]
): PdGreetRuleMetricTriggerInput[] {
  return metrics.filter(isMetricValid).map<PdGreetRuleMetricTriggerInput>((metric) => ({
    code: metric.code!,
    operator: metric.operator!,
    value: metric.value!
  }));
}

function draftAssignmentAsAssignmentInput(
  assignment: AssignTo | OverflowAssignTo
): GroupAssignmentCreateInput | undefined {
  if (assignment.assignmentToType === "None") {
    return undefined;
  }

  if (assignment.assignmentToType === PdGreetAssignmentType.SpecificUserGroup) {
    return {
      assignmentToType: assignment.assignmentToType,
      userGroupId: assignment.userGroupId
    };
  }

  return { assignmentToType: assignment.assignmentToType };
}

export function emptyGreetRuleDraft(siteId: string): GreetRuleDraft {
  return {
    name: "",
    isEnabled: true,
    siteId,
    isIgnoreSuppression: false,
    specialTriggers: {
      sections: null,
      guestType: null,
      tiers: null,
      daysOfWeeks: null
    },
    metricTriggers: [{ __typename: "PdGreetRuleMetricTrigger" }],
    assignment: {
      weight: 0,
      assignTo: null
    }
  };
}

export function convertGreetRuleToDraft(rule: GreetRuleFragment): GreetRuleDraft {
  const { specialTriggers, metricTriggers } = convertTriggersToDraft(rule.triggers);
  const { assignTo, overflowAssignment, overflowAssignment2 } = rule.assignment ?? {};

  const overflow1Allowed =
    !!assignTo?.assignmentToType &&
    assignmentTypeAllowsOverflow(assignTo.assignmentToType);
  const overflow1 = overflow1Allowed
    ? convertOverflowAssignmentToDraft(overflowAssignment ?? null)
    : null;

  const overflow2Allowed =
    !!overflow1?.assignmentToType &&
    assignmentTypeAllowsOverflow(overflow1.assignmentToType);
  const overflow2 = overflow2Allowed
    ? convertOverflowAssignmentToDraft(overflowAssignment2 ?? null)
    : null;

  return {
    id: rule.id,
    name: rule.name,
    isEnabled: rule.isEnabled,
    siteId: rule.site!.id,
    isIgnoreSuppression: rule.isIgnoreSuppression ?? false,
    specialTriggers,
    metricTriggers,
    assignment: {
      weight: rule.assignment?.weight ?? 0,
      assignTo: assignTo ? convertAssignToToDraft(assignTo) : null,
      ...(overflow1 && {
        overflowGroup1: overflow1
      }),
      ...(overflow2 && {
        overflowGroup2: overflow2
      })
    }
  };
}

function triggerGuestType(
  value: PdGreetRuleSpecialTriggerValue
): PdGuestInteractionType | null {
  const guestType = Array.isArray(value.valuesIn) ? value.valuesIn[0] : value.valuesIn;
  if (!guestType) return null;

  switch (guestType) {
    case PdGuestInteractionType.All:
    case PdGuestInteractionType.Coded:
    case PdGuestInteractionType.Uncoded:
      return guestType;
    default:
      return null;
  }
}

function triggerStringMultiselect(
  value: PdGreetRuleSpecialTriggerValue
): DraftMultiSelectValue<string> | null {
  if (value.includeAll) return "__ALL__";

  return Array.isArray(value.valuesIn)
    ? value.valuesIn
        .filter(Boolean)
        .map((val) => (typeof val !== "string" ? `${val}` : val))
    : null;
}

type DraftTriggers = {
  specialTriggers: GreetDraftSiteTriggers;
  metricTriggers: GreetDraftMetricTrigger[];
};

function convertTriggersToDraft(triggers: GreetRuleTrigger[]): DraftTriggers {
  const specialTriggers: GreetDraftSiteTriggers = {
    sections: null,
    guestType: null,
    tiers: null,
    daysOfWeeks: null
  };
  const metricTriggers: GreetDraftMetricTrigger[] = [];

  for (const trigger of triggers) {
    if (trigger.__typename === "PdGreetRuleSpecialTrigger") {
      switch (trigger.type) {
        case "SECTION":
          specialTriggers.sections = triggerStringMultiselect(trigger.specialValue);
          break;
        case "GUEST_TYPE":
          specialTriggers.guestType = triggerGuestType(trigger.specialValue);
          break;
        case "TIER":
          specialTriggers.tiers = triggerStringMultiselect(trigger.specialValue);
          break;
        case "DAYS_OF_WEEKS":
          specialTriggers.daysOfWeeks = triggerStringMultiselect(trigger.specialValue);
          break;
      }
    } else if (trigger.__typename === "PdGreetRuleMetricTrigger") {
      if (!trigger.metric || !isMetricValue(trigger.metricValue)) continue;

      metricTriggers.push({
        __typename: trigger.__typename,
        code: trigger.metric.code,
        operator: trigger.operator,
        valueType: trigger.metric.valueType,
        value: trigger.metricValue
      });
    }
  }

  if (!metricTriggers.length) {
    metricTriggers.push({ __typename: "PdGreetRuleMetricTrigger" });
  }

  return { specialTriggers, metricTriggers };
}

function convertAssignToToDraft(
  group: GreetRuleGroupAssignmentFragment
): AssignTo | null {
  if (!group.assignmentToType) return null;
  return {
    assignmentToType: group.assignmentToType,
    ...(group.userGroup ? { userGroupId: group.userGroup.id } : {})
  };
}

function convertOverflowAssignmentToDraft(
  assignment: GreetRuleGroupAssignmentFragment | null
): OverflowAssignTo | null {
  if (!assignment?.assignmentToType) {
    return { assignmentToType: "None" };
  }

  return convertAssignToToDraft(assignment);
}
