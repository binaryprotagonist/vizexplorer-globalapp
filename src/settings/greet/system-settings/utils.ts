import { GreetReportConfigFragment } from "generated-graphql";
import { NewAlert } from "view-v2/alert";
import { GreetCategory, OptimisticInput } from "./types";
import {
  AboutGreetsChangeParams,
  aboutGreetsSettingTitle,
  aboutGreetsSettings
} from "./about-greets";
import {
  AboutHostsChangeParams,
  aboutHostsSettingTitle,
  aboutHostsSettings
} from "./about-hosts";
import { aboutGuestsSettingTitle, aboutGuestsSettings } from "./about-guests";

const GREET_CATEGORIES: GreetCategory[] = [
  {
    name: "about-guests",
    settings: [...aboutGuestsSettings]
  },
  {
    name: "about-greets",
    settings: [...aboutGreetsSettings]
  },
  {
    name: "about-hosts",
    settings: [...aboutHostsSettings]
  }
];

export const ALERT_SETTING_UPDATED: NewAlert = {
  message: "Setting change saved",
  severity: "success"
};
export const ALERT_SETTING_UPDATE_FAILED: NewAlert = {
  message: "An unexpected error occurred while saving. Please try again.",
  severity: "error"
};

export function reportBannedOptimisticInput(
  newValue: GreetReportConfigFragment
): OptimisticInput {
  return {
    input: { guestReportBanned: newValue },
    newValue: { guestReportBanned: { __typename: "PdGreetReportConfig", ...newValue } }
  };
}

export function aboutGreetsOptimisticInput({
  settingId,
  value
}: AboutGreetsChangeParams): OptimisticInput {
  switch (settingId) {
    case "assign-by-priority-score":
      return {
        input: { greetAssignByPriorityScore: value },
        newValue: { greetAssignByPriorityScore: value }
      };
    case "suppression-days-after-completion":
      return {
        input: { greetSuppressionDays: { coded: value.coded, uncoded: value.uncoded } },
        newValue: { greetSuppressionDays: value }
      };
    case "queue-inactive-greet-timeout":
      return {
        input: {
          greetQueueInactiveTimeout: { hours: value.hours, minutes: value.minutes }
        },
        newValue: { greetQueueInactiveTimeout: value }
      };
    case "greet-reassignment-timeout":
      return {
        input: {
          greetReassignmentTimeout: { hours: value.hours, minutes: value.minutes }
        },
        newValue: { greetReassignmentTimeout: value }
      };

    case "show-guests-active-actions":
      return {
        input: { greetShowGuestActiveActions: value },
        newValue: { greetShowGuestActiveActions: value }
      };
  }
}

export function aboutHostsOptimisticInput({
  settingId,
  value
}: AboutHostsChangeParams): OptimisticInput {
  switch (settingId) {
    case "enable-section-for-greet":
      return {
        input: { hostEnableSections: value },
        newValue: { hostEnableSections: value }
      };
    case "allow-suppression-without-completion":
      return {
        input: { hostAllowSuppression: value },
        newValue: { hostAllowSuppression: value }
      };
    case "max-assignment-per-host":
      return {
        input: { hostMaxAssignments: value },
        newValue: { hostMaxAssignments: value }
      };
    case "max-missed-greets": {
      return {
        input: { hostMaxMissedGreets: value },
        newValue: { hostMaxMissedGreets: value }
      };
    }
  }
}

function filterCategory(category: GreetCategory, search: string): GreetCategory | null {
  if (category.name === "about-guests") {
    const settings = category.settings.filter((setting) => {
      const title = aboutGuestsSettingTitle(setting);
      return title.toLowerCase().includes(search);
    });
    return settings.length ? { name: "about-guests", settings } : null;
  }

  if (category.name === "about-greets") {
    const settings = category.settings.filter((setting) => {
      const title = aboutGreetsSettingTitle(setting);
      return title.toLowerCase().includes(search);
    });
    return settings.length ? { name: "about-greets", settings } : null;
  }

  if (category.name === "about-hosts") {
    const settings = category.settings.filter((setting) => {
      const title = aboutHostsSettingTitle(setting);
      return title.toLowerCase().includes(search);
    });
    return settings.length ? { name: "about-hosts", settings } : null;
  }

  return null;
}

export function filteredGreetCategories(search: string): GreetCategory[] {
  if (!search.length) return GREET_CATEGORIES;

  const normalizedSearch = search.trim().toLowerCase();
  const categories: GreetCategory[] = [];
  for (const category of GREET_CATEGORIES) {
    const filteredCategory = filterCategory(category, normalizedSearch);
    if (filteredCategory) {
      categories.push(filteredCategory);
    }
  }

  return categories;
}
