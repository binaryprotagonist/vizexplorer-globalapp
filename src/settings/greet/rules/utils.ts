import {
  GaUserFragment,
  GreetRuleFragment,
  GreetRuleMetricTriggerFragment,
  GreetRuleSpecialTriggerFragment,
  PdGreetAssignmentType,
  PdGreetMetricValueType,
  PdGreetRuleConditionOperator,
  PdGreetRuleSpecialTriggerValue,
  PdGreetRuleTriggerType,
  PdGuestInteractionType,
  Scalars,
  SiteFragment
} from "generated-graphql";
import {
  GreetRuleTrigger,
  METRIC_DATE_OPTIONS,
  MetricDateOption,
  MetricValue,
  SpecialTriggerObj
} from "./types";
import { canUser, isOrgAdmin } from "../../../view/user/utils";
import { UserActionType } from "../../../view/user/types";

export function ruleMatchesSearch(rule: GreetRuleFragment, search: string): boolean {
  return rule.name.toLowerCase().includes(search.toLowerCase());
}

export function isSpecialTrigger(
  trigger: GreetRuleTrigger
): trigger is GreetRuleSpecialTriggerFragment {
  return trigger.__typename === "PdGreetRuleSpecialTrigger";
}

export function isMetricTrigger(
  trigger: GreetRuleTrigger
): trigger is GreetRuleMetricTriggerFragment {
  return trigger.__typename === "PdGreetRuleMetricTrigger";
}

export function buildSpecialTriggerObj(
  triggers: GreetRuleSpecialTriggerFragment[]
): SpecialTriggerObj {
  return triggers.reduce<SpecialTriggerObj>((acc, cur) => {
    switch (cur.type) {
      case PdGreetRuleTriggerType.Section:
        acc.sections = cur.specialValue;
        break;
      case PdGreetRuleTriggerType.GuestType:
        acc.guestType = cur.specialValue;
        break;
      case PdGreetRuleTriggerType.Tier:
        acc.tiers = cur.specialValue;
        break;
      case PdGreetRuleTriggerType.DaysOfWeeks:
        acc.daysOfWeeks = cur.specialValue;
        break;
    }

    return acc;
  }, {});
}

export function greetAssignmentTypeLabel(type: PdGreetAssignmentType): string {
  switch (type) {
    case PdGreetAssignmentType.GuestHost:
      return "Guest Host Only";
    case PdGreetAssignmentType.GuestHostUserGroup:
      return "Guest Coded User Group";
    case PdGreetAssignmentType.SpecificUserGroup:
      return "Specific User Group";
    case PdGreetAssignmentType.AllUsers:
      return "All Hosts";
  }
}

export function specialTriggerValueLabel(
  value: PdGreetRuleSpecialTriggerValue,
  customAllText: string
): string {
  if (value.includeAll) return customAllText;
  if (!value.valuesIn) return "Invalid value";

  if (!Array.isArray(value.valuesIn)) {
    return `${value.valuesIn}`;
  }

  return value.valuesIn.join(", ");
}

export function guestTypeValueLabel(value: PdGreetRuleSpecialTriggerValue): string {
  const guestType = Array.isArray(value.valuesIn) ? value.valuesIn[0] : value.valuesIn;
  if (!guestType) return "Unknown";

  switch (guestType) {
    case PdGuestInteractionType.All:
      return "All guests";
    case PdGuestInteractionType.Coded:
      return "Coded";
    case PdGuestInteractionType.Uncoded:
      return "Uncoded";
    default:
      return `${guestType}`;
  }
}

export function metricTriggerValueLabel(trigger: GreetRuleMetricTriggerFragment): string {
  switch (trigger.metric?.valueType) {
    case PdGreetMetricValueType.Numeric:
      return numericTriggerValueLabel(trigger);
    case PdGreetMetricValueType.Date:
      return dateMetricValueLabel(trigger);
    default:
      return "Unknown metric type";
  }
}

function numericTriggerValueLabel(trigger: GreetRuleMetricTriggerFragment): string {
  const value = trigger.metricValue;

  switch (trigger.operator) {
    case PdGreetRuleConditionOperator.Range: {
      const range = Array.isArray(value) ? value : [];
      const min = range[0] ?? null;
      const max = range[1] ?? null;

      if (min === null && max === null) {
        return "Invalid range value";
      }

      const minLabel = `Min ${min}`;
      const maxLabel = `Max ${max}`;
      return min !== null && max !== null
        ? `${minLabel} ${maxLabel}`
        : min !== null
          ? minLabel
          : maxLabel;
    }
    case PdGreetRuleConditionOperator.Equal: {
      const eqValue = Array.isArray(value) ? value[0] : value;
      return eqValue ? `${eqValue}` : "Invalid value";
    }
    case PdGreetRuleConditionOperator.GreaterOrEqual: {
      const geValue = Array.isArray(value) ? value[0] : value;
      return geValue ? `Min ${geValue}` : "Invalid value";
    }
    case PdGreetRuleConditionOperator.LessOrEqual: {
      const leValue = Array.isArray(value) ? value[0] : value;
      return leValue ? `Max ${leValue}` : "Invalid value";
    }
    case PdGreetRuleConditionOperator.In: {
      if (!Array.isArray(value) || !value.length) {
        return "Invalid value";
      }
      return value.join(", ");
    }
  }
}

function dateMetricValueLabel(trigger: GreetRuleMetricTriggerFragment): string {
  const value = Array.isArray(trigger.metricValue) ? trigger.metricValue : [];

  if (isMetricValue(value)) {
    return findMetricDateByRawValue(value)?.label ?? "Invalid date value";
  }

  return "Invalid date value";
}

export function findSpecialTrigger(
  triggers: GreetRuleTrigger[],
  type: PdGreetRuleTriggerType
): GreetRuleSpecialTriggerFragment | undefined {
  return triggers.find(
    (trigger) => isSpecialTrigger(trigger) && trigger.type === type
  ) as GreetRuleSpecialTriggerFragment | undefined;
}

export function sitesManageableByUser(sites: SiteFragment[], user: GaUserFragment) {
  if (isOrgAdmin(user.accessLevel)) {
    return sites;
  }

  return sites.filter((site) =>
    canUser(user, { type: UserActionType.MANAGE_GREET_RULES, siteId: site.id })
  );
}

export function findMetricDateByRawValue(
  rawValue: MetricValue
): MetricDateOption | undefined {
  if (typeof rawValue[0] !== "number" || typeof rawValue[1] !== "number") return;

  return METRIC_DATE_OPTIONS.find((option) => {
    return option.rawValue[0] === rawValue[0] && option.rawValue[1] === rawValue[1];
  });
}

export function isMetricValue(
  value: Scalars["PdGreetRuleConditionValue"]["input"]
): value is MetricValue {
  if (!Array.isArray(value)) return false;
  if (value.length !== 2) return false;
  if (typeof value[0] !== "number" && value[0] !== null) return false;
  if (typeof value[1] !== "number" && value[1] !== null) return false;
  return true;
}
