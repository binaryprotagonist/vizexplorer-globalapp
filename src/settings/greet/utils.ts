import { GaUserFragment } from "generated-graphql";
import { GreetView } from "./types";
import { isOrgAdmin } from "../../view/user/utils";

export function greetSettingsTabLabel(type: GreetView): string {
  switch (type) {
    case "greet-rules":
      return "Greet rules";
    case "system-settings":
      return "System settings";
  }
}

export function canAccessGreetView(view: GreetView, user: GaUserFragment): boolean {
  switch (view) {
    case "greet-rules":
      return true;
    case "system-settings":
      return isOrgAdmin(user.accessLevel);
  }
}
