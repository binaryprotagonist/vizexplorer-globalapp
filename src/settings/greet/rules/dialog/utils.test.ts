import { mockGreetRuleWithMultiselectedOptions } from "testing/mocks";
import { convertGreetRuleToDraft } from "./utils";
import { produce } from "immer";
import { PdGreetAssignmentType } from "generated-graphql";

describe("Greet Rule Dialog utils", () => {
  describe("convertGreetRuleToDraft", () => {
    const mockGreetRule = mockGreetRuleWithMultiselectedOptions;

    it("converts a rule to the expected draft format", () => {
      expect(convertGreetRuleToDraft(mockGreetRule)).toMatchSnapshot();
    });

    it("converts overflow1 to None if overflow1 is null and assignTo is defined", () => {
      const nullOverflow1 = produce(mockGreetRule, (draft) => {
        draft.assignment!.overflowAssignment = null;
        draft.assignment!.overflowAssignment2 = null;
      });
      const convertedRule = convertGreetRuleToDraft(nullOverflow1);

      expect(convertedRule.assignment.overflowGroup1!.assignmentToType).toEqual("None");
      expect(convertedRule.assignment.overflowGroup2).toBeUndefined();
    });

    it("doesn't convert overflow1 to None if assignTo assignment type is All Users", () => {
      const nullOverflow1 = produce(mockGreetRule, (draft) => {
        draft.assignment!.assignTo!.assignmentToType = PdGreetAssignmentType.AllUsers;
        draft.assignment!.overflowAssignment = null;
        draft.assignment!.overflowAssignment2 = null;
      });
      const convertedRule = convertGreetRuleToDraft(nullOverflow1);

      expect(convertedRule.assignment.overflowGroup1).toBeUndefined();
      expect(convertedRule.assignment.overflowGroup2).toBeUndefined();
    });

    it("converts overflow2 to None if overflow2 is null and overflow1 is defined", () => {
      const nullOverflow2 = produce(mockGreetRule, (draft) => {
        draft.assignment!.overflowAssignment2 = null;
      });
      const convertedRule = convertGreetRuleToDraft(nullOverflow2);

      expect(convertedRule.assignment.overflowGroup2!.assignmentToType).toEqual("None");
    });

    it("converts overflow2 to None if overflow2 assignment type is null and overflow1 is defined", () => {
      const nullOverflow2 = produce(mockGreetRule, (draft) => {
        draft.assignment!.overflowAssignment2!.assignmentToType = null;
      });
      const convertedRule = convertGreetRuleToDraft(nullOverflow2);

      expect(convertedRule.assignment.overflowGroup2!.assignmentToType).toEqual("None");
    });

    it("doesn't convert overflow2 to None if overflow1 is null", () => {
      const nullOverflow2 = produce(mockGreetRule, (draft) => {
        draft.assignment!.overflowAssignment = null;
        draft.assignment!.overflowAssignment2 = null;
      });
      const convertedRule = convertGreetRuleToDraft(nullOverflow2);

      expect(convertedRule.assignment.overflowGroup2).toBeUndefined();
    });

    it("doesn't convert overflow2 to None if overflow1 assignment type is All Users", () => {
      const nullOverflow2 = produce(mockGreetRule, (draft) => {
        draft.assignment!.overflowAssignment!.assignmentToType =
          PdGreetAssignmentType.AllUsers;
        draft.assignment!.overflowAssignment2 = null;
      });
      const convertedRule = convertGreetRuleToDraft(nullOverflow2);

      expect(convertedRule.assignment.overflowGroup2).toBeUndefined();
    });
  });
});
