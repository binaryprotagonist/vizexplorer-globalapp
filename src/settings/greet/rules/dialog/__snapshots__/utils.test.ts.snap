// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Greet Rule Dialog utils convertGreetRuleToDraft converts a rule to the expected draft format 1`] = `
{
  "assignment": {
    "assignTo": {
      "assignmentToType": "SPECIFIC_USER_GROUP",
      "userGroupId": "39",
    },
    "overflowGroup1": {
      "assignmentToType": "SPECIFIC_USER_GROUP",
      "userGroupId": "40",
    },
    "overflowGroup2": {
      "assignmentToType": "SPECIFIC_USER_GROUP",
      "userGroupId": "47",
    },
    "weight": 75,
  },
  "id": "1",
  "isEnabled": true,
  "isIgnoreSuppression": false,
  "metricTriggers": [
    {
      "__typename": "PdGreetRuleMetricTrigger",
      "code": "todayActualWin",
      "operator": "RANGE",
      "value": [
        100,
        200,
      ],
      "valueType": "NUMERIC",
    },
  ],
  "name": "JL Test",
  "siteId": "0",
  "specialTriggers": {
    "daysOfWeeks": [
      "Sunday",
      "Tuesday",
      "Thursday",
      "Saturday",
    ],
    "guestType": "ALL",
    "sections": [
      "01",
      "02",
    ],
    "tiers": [
      "Platinum",
      "Gold",
    ],
  },
}
`;
